### ISSUE TITLE
jumpboots broken

### ISSUE THREAD

**Issue by Tlaltecuhtli** [tgstation/tgstation#34432]
Issue reported from Round ID: 81883 (/tg/Station Terry [EU] [100% LAG FREE])

jump boots stop working after a few uses

**Comment by Catistian** [comment]
yep, I can attest to this. they stop working entirely, even buying a new pair won't make them work again

**Comment by WarRage333** [comment]
This useless item is still broke, stops working after 4 uses.

**Comment by KomradeSpectre** [comment]
Additional Info on this issue, and the key to why they break:

- Jumpboots break if you run into a wall midjump with them, or get stunned while jumping. Anything that interrupts the resolution of the `jumping =` check.

**Comment by Caiggas** [comment]
Update: I may have to set this aside for a bit, as I am tired. Here is what I have uncovered so far:

The boots break if you run into anything that stops the jump WITHIN 2 tiles. Anything further than that, so far, has not broken the boots. Buying a new pair DOES work, as the "jumping" check is per instance of the boots. I could not figure out how to test being stunned while jumping in my test server.

So when the boots work like they are supposed to, the following happens:
1: User presses jump button
2: The boots check to see if the user is alive, the jumping var is false, and recharging is complete.
3: If these conditions are met, the boots call "throw_at" with some stats from the boots and a callback to "hop_end"
4: throw_at checks a few conditions to make sure the user can be thrown, passes "true" to the boots, then throws the user per the information given to them by the boots.
5: The "throw_at" call in the boots was in an if statement, which when it receives the "true" from the call, it sets "jumping" to true, plays a sound, and logs the jump in chat.
6: When the throw is complete, it runs the callback to "hop_end".
7: "hop_end" sets the "jumping" var to false and starts the recharge timer.

Colliding with any impassible tile or getting knocked over within 2 tiles of your starting point appears to make the "jumping" var never get set back to "false". As far as I can figure, this MUST be caused by EITHER the callback or "hop_end" getting interrupted somehow. I don't know why interrupting the throw causes this to occur, or why it only occurs in the first 2 tiles. 
My hunch is that if the throw is interrupted within a certain AMOUNT OF TIME, it will break, and that happens to coincide with about 2 tiles. This makes me think the culprit is that the callback never occurs if the throw is interrupted quickly enough, but I need to test further.

Note that the callback explicitly is NOT supposed to occur if the "throw_at" call returns false, but that is NOT the culprit here, as the throw NEVER occurs at all in that case. Instead, there is some flavor text that plays to let the user know the jump failed.

**Comment by Caiggas** [comment]
btw, the relavent code for this is at the following paths.
Boots:
tgstation/code/modules/clothing/shoes/miscellaneous.dm
Throw_at:
tgstation/code/game/atoms_movable.dm
callbacks:
tgstation/code/datums/callback.dm

**Comment by Caiggas** [comment]
Alright, I isolated the bug. So I'm gonna do my best to describe this clearly. The issue is NOT that the hop_end function never gets called. the issue is that the hop_end function gets called BEFORE the "jumping" var ever gets set to true. 

So to reiterate, normally you press the button, the jump calculates and you get "jumping" set to true. Then your character jumps, hop_end gets called, and "jumping" gets set to false.
When the boots break, the jump calculates, occurs, hop_end gets called, "jumping" gets set to false, then jumping gets set right back to true and never gets set back to false.

Looking at that description it doesn't make much sense, but if you look at the code it will fall together. I am mostly making this comment because I need to go to bed and I don't want to forget this for when I fix it tomorrow. Note I verified this with a large number of debug lines that tell me when each involved function gets called, and when jumping gets changed.
